### PR TITLE
drivers: gpio: Add fxl6408 driver

### PR DESCRIPTION
**Pull request by Ilagas**
Add driver for the fxl6408 gpio-expander. The driver makes it possible to read and write to the gpio pins over I2C. See the README file of the sample application for general usage of the driver. The driver is inspired by the PCA95XX driver. Currently, the driver does not support interrupts or HIGH-Z mode which are options on the fxl6408. I'm not completely sure about HIGH-Z mode; should HIGH-Z mode be included in the driver and if so, how? In the gpio api, I could not find any support for HIGH-Z.

Signed-off-by: Imre Lagas Imre_Lagas@hotmail.com

### CODE REVIEW

File: drivers/gpio/Kconfig.fxl6408 [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# ST Microelectronics STM32L151XC MCU
```
Comment (galak) [inline drivers/gpio/Kconfig.fxl6408:1]: This comment about STM32L151XC seems out of place.

File: dts/bindings/gpio/fcs,fxl6408.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+compatible: "fcs,fxl6408"
+
+include: [gpio-controller.yaml, i2c-device.yaml]
+
+properties:
+    label:
+      required: true
+
+    has-pud:
```
Comment (galak) [inline dts/bindings/gpio/fcs,fxl6408.yaml:14]: Where is this property being used in the code?

**Review by galak — COMMENTED** [review]
Why do we need a unique sample for this driver?

**Review by galak — CHANGES_REQUESTED** [review]
please rebase to remove the merge commit.

File: drivers/gpio/Kconfig.fxl6408 (hunk 1 above)
Comment (Ilagas) [reply drivers/gpio/Kconfig.fxl6408]: I've removed this line

File: dts/bindings/gpio/fcs,fxl6408.yaml (hunk 2 above)
Comment (Ilagas) [reply dts/bindings/gpio/fcs,fxl6408.yaml]: I've removed the has-pud part since it wasn't used in code.

**Comment by Ilagas** [comment]
@galak, I added the sample application mainly because i thought it would be useful for verifying that the driver is working. Should I remove the sample application? I've also pushed a new commit which I believe should resolve the merge commit.

**Comment by Ilagas** [comment]
I have deleted the sample application since it didn't add that much and i had included it more or less to show that the driver is working. Just for completeness; I've tested all functions in the gpio API and they all work.

File: dts/bindings/gpio/fcs,fxl6408.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2021 Abel Sensors
+# SPDX-License-Identifier: Apache-2.0
+
+description: FXL6408 I2C-based GPIO expander
+
+compatible: "fcs,fxl6408"
```
Comment (galak) [inline dts/bindings/gpio/fcs,fxl6408.yaml:6]: Curious where `fcs` is coming from?

**Review by galak — CHANGES_REQUESTED** [review]
In general its best to merge commits and force push updates to your PR when you make changes so the commits are just the logical commits that would end up getting merged. 

So for example, since you are dropping the sample, the commit change that adds and and removes it should go away.

Additionally, all the other changes, dts, initial driver, and fixes for code compliance should get merged into one commit.

File: dts/bindings/gpio/fcs,fxl6408.yaml (hunk 3 above)
Comment (Ilagas) [reply dts/bindings/gpio/fcs,fxl6408.yaml:6]: I found it in /zephyr/dts/bindings/vendor-prefixes.txt. It stands for the vendor Fairchild Semiconductor.

File: drivers/gpio/gpio_fxl6408.c [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+	ret = update_output_regs(dev, reg_out);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = update_direction_regs(dev, reg_dir);
+	return ret;
+}
+
+int gpio_fxl6408_init(const struct device *device)
```
Comment (mnkp) [inline drivers/gpio/gpio_fxl6408.c:211]: We traditionally keep _init function toward the bottom of the file. Just above the `struct` defining API functions. Let's move it there.

**Review by mnkp — COMMENTED** [review]
This is a partial review.

> should HIGH-Z mode be included in the driver and if so, how? In the gpio api, I could not find any support for HIGH-Z.

The driver should configure pin in High-Z state if neither `GPIO_INPUT` nor `GPIO_OUTPUT` flag is provided.

**Comment by Ilagas** [comment]
I've run into a problem. I have rebased to master, but now whenever I try to build I get the following warning:
```
/home/ilagas/zephyrproject/zephyr/drivers/gpio/gpio_fxl6408.c:436:13: warning: Macro is deprecated
  436 | DT_INST_FOREACH_STATUS_OKAY(GPIO_FXL6408_DEVICE_INSTANCE)
```
I've found this slack thread which states that it might be a bug: https://zephyrproject.slack.com/archives/C18PLHN8H/p1620392150084400

I dont really know where to start looking to fix this warning. Can I continue the PR or should this issue be fixed first?

**Comment by mnkp** [comment]
> ```
> /home/ilagas/zephyrproject/zephyr/drivers/gpio/gpio_fxl6408.c:436:13: warning: Macro is deprecated
>   436 | DT_INST_FOREACH_STATUS_OKAY(GPIO_FXL6408_DEVICE_INSTANCE)
> ```

Unfortunately, the message is confusing but there is probably no way to avoid it. The power management has recently been reworked and `device_pm_control_nop` function was removed. You need to replace `device_pm_control_nop` with `NULL` in `DEVICE_DT_INST_DEFINE` macro.

File: dts/bindings/gpio/fcs,fxl6408.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+include: [gpio-controller.yaml, i2c-device.yaml]
+
+properties:
+    label:
+      required: true
+
+    "#gpio-cells":
+      const: 1
```
[the remark this answers is not in the document]
Comment (Ilagas) [reply dts/bindings/gpio/fcs,fxl6408.yaml]: I've changed this to `const: 2`. I must have misinterpreted this value; what is this used for exactly?

[the remark this answers is not in the document]
Comment (mnkp) [reply dts/bindings/gpio/fcs,fxl6408.yaml]: This is indeed not very well documented in DT Specification. This is related to 
```
gpio-cells:
  - pin
  - flags
```
and tells how many parameters are passed after the gpio phandle in `*gpios` parameter, as in 
```
gpios = <&pioc 8 GPIO_ACTIVE_LOW>;
```

File: drivers/gpio/gpio_fxl6408.c [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+
+	ret = i2c_reg_write_byte(i2c_master, config->i2c_slave_addr, reg,
+					 value);
+	if (ret == 0) {
+		*cache = value;
+	} else {
+		LOG_ERR("error writing to register 0x%X (%d)", reg,
+			ret);
+	}
+	LOG_DBG("Write: REG[0x%X] = 0x%X", reg, *cache);
```
[the remark this answers is not in the document]
Comment (Ilagas) [reply drivers/gpio/gpio_fxl6408.c:116]: What do you mean by this exactly? Do you mean that `read_port_regs` returns ret if it fails? If so, should both return out of the function on fail or should neither?

**Comment by Ilagas** [comment]
It seems I have broken things with my latest push. Ill fix it as soon as possible.

**Comment by Ilagas** [comment]
I just noticed that the buildkite test has failed. When I look into it, it has problems with nxp,lpc-gpio.yaml. I did not change anything in this file and I don't get why this different board would be causing problems. Any ideas why this is happening?

**Comment by PerMac** [comment]
> I just noticed that the buildkite test has failed. When I look into it, it has problems with nxp,lpc-gpio.yaml. I did not change anything in this file and I don't get why this different board would be causing problems. Any ideas why this is happening?

@Ilagas Check the "Files changed" by your PR. You have changed this file in fact. The buildkit error is not a CI issue.

File: dts/bindings/gpio/nxp,lpc-gpio.yaml [hunk 7]
```diff
@@ -17,14 +17,6 @@ properties:
...
       const: 2
 
-    port:
-      type: int
-      required: true
-      description: The GPIO port this node describes
-      enum:
-        - 0
-        - 1
-
```
Comment (PerMac) [inline dts/bindings/gpio/nxp,lpc-gpio.yaml:27]: @Ilagas I guess this relates to the error you see in buildkite

File: drivers/gpio/gpio_fxl6408.c [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+	}
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	ret = setup_pin_dir(dev, pin, flags);
+	if (ret) {
```
Comment (mnkp) [inline drivers/gpio/gpio_fxl6408.c:275]: Let's write it as `(ret != 0)`, here and in a few more places, to use the same style. This is recommended (but not yet enforced) but the coding guidelines.

File: drivers/gpio/gpio_fxl6408.c [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+			.pud_en = 0xFF,                                                 \
+			.pud_sel = 0xFF                                                 \
+		}                                                                   \
+	};                                                                      \
+\
+	DEVICE_DT_INST_DEFINE(inst, gpio_fxl6408_init, NULL,                    \
+				  &gpio_fxl6408_##inst##_drvdata,                           \
+				  &gpio_fxl6408_##inst##_cfg, POST_KERNEL,                  \
+				  CONFIG_GPIO_FXL6408_INIT_PRIORITY,                        \
+				  &gpio_fxl_driver);
```
Comment (mnkp) [inline drivers/gpio/gpio_fxl6408.c:436]: Let's limit line length to 80 characters.

**Review by mnkp — COMMENTED** [review]
LGTM with a few more, mainly cosmetic changes. @Ilagas could you please test the driver with `tests/drivers/gpio/gpio_api_1pin/` testcase.

File: dts/bindings/gpio/fcs,fxl6408.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (mnkp) [reply dts/bindings/gpio/fcs,fxl6408.yaml:15]: It looks like it has been reverted back to 1. Should be `const: 2`.

**Review by mnkp — CHANGES_REQUESTED** [review]
@Ilagas please confirm that tests/drivers/gpio/gpio_api_1pin/ testcase is passing.

Some CI checks are failing. Please clean it up.

**Comment by Ilagas** [comment]
> @Ilagas please confirm that tests/drivers/gpio/gpio_api_1pin/ testcase is passing.
> 
> Some CI checks are failing. Please clean it up.

The gpio_api_1pin test runs succesfully. I've tested it on a stm32  with the  FXL6408.  I've used the following line to run the test:
`scripts/twister -vv --device-testing --device-serial /dev/ttyUSB0 -p stm32l1_disco -T tests/drivers/gpio/gpio_api_1pin/`

The twister.log file should be attached. 
[twister.log](https://github.com/zephyrproject-rtos/zephyr/files/6704262/twister.log)

**Comment by Ilagas** [comment]
What does strike me as odd though is that when running the  tests/drivers/gpio/gpio_api_1pin/ testcase, I would assume the LEDs attached to the FXL6408 would be blinking. This is not the case. Is this a problem?

**Comment by Ilagas** [comment]
> What does strike me as odd though is that when running the tests/drivers/gpio/gpio_api_1pin/ testcase, I would assume the LEDs attached to the FXL6408 would be blinking. This is not the case. Is this a problem?

Whenever I run the test, it doesn't test the gpio-expander but rather the GPIO pins of the stm32 itself. Any idea how I can configure the testcase to test the pins of the FXL6408?

**Comment by Ilagas** [comment]
By the way, a while back I added accidentally comitted a few unrelated files which trigerred Github to add the NXP platfrom tag. This added a few additional reviewers to the PR. Should the tag and the reviewers be manually removed or does it not matter that much?

**Review by mnkp — COMMENTED** [review]
> There seems to be a problem. Whenever I run the test, it doesn't test the gpio-expander but rather the GPIO pins of the stm32 itself. Any idea how I can configure the testcase to test the pins of the FXL6408? I need to change led0 in the device tree so that is uses a pin from the FXL6408, but I cant seem to make this work.

To test is quickly on stm32l1_disco board you can simply modify `gpios` property of the `green_led` / `led0` to use FXL6408 node that you already have. If unsure you can check `boards/` folder for existing examples. Other in tree GPIO-expanders are cy8c95xx, pca95xx, pcal6408a, sx1509b. If you grep board directory for those names you'll definitely find something.

> By the way, a while back I added accidentally comitted a few unrelated files which trigerred Github to add the NXP platfrom tag. This added a few additional reviewers to the PR. Should the tag and the reviewers be manually removed from the PR or does it not matter that much?

Normally you would remove reviewers that were added accidentally. In case of this PR, since there were few, you can probably leave it as is.

**Comment by Ilagas** [comment]
> @Ilagas please confirm that tests/drivers/gpio/gpio_api_1pin/ testcase is passing.

I've tried running this test but many of the tests are failing because they require the pin to be configured as both input and output. I don't think it's possible on the fxl6408 to write to a pin and read it at the same time. I've come to this conclusion by trying to write/read a pin in a sample application. It also states in the datasheet that when a pin is configured as output the input status register bit of that pin has no effect (see table 10 in the datasheet). For now I'm removing the functionality to set both input and output for a pin but I think it would be wise of someone could doublecheck if this really is not possoible on the fxl6408.

**Comment by Ilagas** [comment]
I've succesfully ran the tests/drivers/gpio/gpio_api_1pin/ testcase. Do note that many of the tests were simply skipped because setting input and ouput at the same time was not supported. 
[twister.log](https://github.com/zephyrproject-rtos/zephyr/files/6732992/twister.log)

File: drivers/gpio/gpio_fxl6408.c [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+}
+
+static inline int update_pul_en_regs(const struct device *dev, uint8_t value)
+{
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+
+	return write_port_regs(dev, REG_PUD_EN, &drv_data->reg_cache.pud_en,
+				   value);
+}
```
[the remark this answers is not in the document]
Comment (mnkp) [reply drivers/gpio/gpio_fxl6408.c]: Sorry, I put the comment in the wrong place. That was misleading. It's easier / cleaner to implement this check inside `write_port_regs` function.

**Comment by mnkp** [comment]
> I've succesfully ran the tests/drivers/gpio/gpio_api_1pin/ testcase. Do note that many of the tests were simply skipped because setting input and ouput at the same time was not supported.

I double checked the datasheet and indeed in/out mode is not supported. The gpio_api_1pin testcase is not very useful in this case but at least the `gpio_pin_configure` function returns now the correct return value.

@Ilagas in that case could you test the driver with `tests/drivers/gpio/gpio_basic_api/` testcase. It's a bit more cumbersome to setup but you'll find multiple examples in the `tests/drivers/gpio/gpio_basic_api/boards/` directory.

@galak I'll need your help here. I assume that fxl6408 gpio-expander is not available on any of the evaluation boards. Do we want to have in-tree board overlay/config files to enable testing the driver on some arbitrary board with `tests/drivers/gpio/gpio_basic_api/` testcase or is it enough that we add an entry to `tests/drivers/build_all/gpio/` to make sure the driver builds?

**Comment by Ilagas** [comment]
I finally managed to get the test `tests/drivers/gpio/gpio_basic_api/` up and running. I've used one of the supported boards; the `nucleo_g071rb`. The test is failing however because the test tries to set the `GPIO_ACTIVE_LOW` flag. This is currently not supported. I cant find any other implementation of the `GPIO_ACTIVE_LOW` flag and some other drivers don´t support it. If you would like to implement the `GPIO_ACTIVE_LOW` flag, what would be the best way to do this?

**Comment by mnkp** [comment]
> The test is failing however because the test tries to set the `GPIO_ACTIVE_LOW` flag.

The flag is handled fully by the code in `include/drivers/gpio.h`. There is nothing driver needs to do apart from making sure that `struct gpio_driver_data common;` is the first field in the runtime data struct, which it is. The testcase must be failing for some other reason.

**Comment by Ilagas** [comment]
I have ran the `gpio_basic_api` testcast testcase partially succesful. The failure I encountered earlier was because the `GPIO_ACTIVE_LOW` flag was not in my list of supported flags. As far as I understand I can just add this flag to the list of supported flags so I've done this. 
The test that writes and reads the pins executes succesful and many of the tests to test callbacks are skipped because gpio interrupts are not supported by the driver. Only the last test fails because It tries to configure the pin with the flag `GPIO_INT_DEBOUNCE`. This is not supported by the driver and it's not in the list of supported flags. How should this be adressed?

[handler.log](https://github.com/zephyrproject-rtos/zephyr/files/6825595/handler.log)

**Comment by Ilagas** [comment]
I just notices many of the buildkite tests have failed. This time I didn't accidentally include unrelated files or made changes in any unrelated files. Also, I performed a rebase to the main branch before pushing. I have no idea why these unrelated errors pop up. Any ideas what could be causing this?

**Comment by mnkp** [comment]
> Only the last test fails because It tries to configure the pin with the flag `GPIO_INT_DEBOUNCE`. This is not supported by the driver and it's not in the list of supported flags. How should this be adressed?

Yes,  `GPIO_INT_DEBOUNCE` is a bit special. According to the documentation

```
 * @note Drivers that do not support a debounce feature should ignore
 * this flag rather than rejecting the configuration with -ENOTSUP.
```

So you need to add the flag to the list of supported flags. Even if it's not supported by the hardware.

**Comment by mnkp** [comment]
@Ilagas there is one more change required. You'll need to add an entry to the `tests/drivers/build_all/gpio/` testcase to make sure the driver is build by the CI. It's not complicated. You can `git grep -i PCAL6408A tests/drivers/build_all/gpio/` to see what needs to be done. PCAL6408A is another GPIO expander.

Also, please rebase the PR on the latest master.

**Comment by Ilagas** [comment]
i have added the testcase to `tests/drivers/build_all/gpio/` and added `GPIO_INT_DEBOUNCE` in the list of allowed tags. The test runs succesfully now:

[twister.log](https://github.com/zephyrproject-rtos/zephyr/files/6840479/twister.log)

**Review by mnkp — COMMENTED** [review]
LGTM. For the build_all testcase you need to enable the driver by setting `CONFIG_GPIO_FXL6408=y` in `tests/drivers/build_all/gpio/prj.conf`.

**Comment by Ilagas** [comment]
I've added `CONFIG_GPIO_FXL6408=y` to the prj.conf file. I have added it in the middle so it's underneath the `CONFIG_I2C` line and not underneath `CONFIG_SPI`. Are there any rules as to where to add lines or should I just always add lines to the end of a file?

**Review by mnkp — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Comment by carlescufi** [comment]
@Ilagas could you please rebsae this?

**Comment by Ilagas** [comment]
I have rebased to the zephyr main branch

**Comment by Ilagas** [comment]
Hi, do I need to do anything additional? If possible I would like to finish the PR soon.

**Comment by PerMac** [comment]
This PR is out of my competence scope and I removed myself from the reviewers. Github shows that I had a comment, but it is no longer applicable as that was resolved.

**Pull request merged**